src/Routing/Next.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace Tnapf\Router\Routing;
+
+use LogicException;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use stdClass;
+use Tnapf\Router\Interfaces\RequestHandlerInterface;
+
+class Next {
+    /**
+     * @var class-string<RequestHandlerInterface>[]
+     */
+    protected array $middleware = [];
+
+    /**
+     * @var class-string<RequestHandlerInterface>[]
+     */
+    protected array $postware = [];
+    private bool $completed = false;
+    private int $middlewarePosition = 0;
+    private int $postwarePosition = 0;
+
+    /**
+     * @param class-string<RequestHandlerInterface> $requestHandler
+     */
+    public function __construct(protected readonly string $requestHandler) {}
+
+    /**
+     * @var class-string<RequestHandlerInterface>[]
+     */
+    public function addMiddleware(string ...$middleware): self
+    {
+        if ($this->completed) {
+            throw new LogicException("This instance has already been marked complete and cannot have anymore middleware added");
+        }
+
+        foreach ($middleware as $middlewareClass) {
+            $this->middleware[] = $middlewareClass;
+        }
+
+        return $this;
+    }
+
+    /**
+     * @var class-string<RequestHandlerInterface>[]
+     */
+    public function addPostware(string ...$postware): self
+    {
+        if ($this->completed) {
+            throw new LogicException("This instance has already been marked complete and cannot have anymore postware added");
+        }
+
+        foreach ($postware as $postwareClass) {
+            $this->postware[] = $postwareClass;
+        }
+
+        return $this;
+    }
+
+    public function markComplete(): self
+    {
+        if (!$this->completed) {
+            $this->completed = false;
+        }
+
+        $this->middleware[] = $this->requestHandler;
+
+        return $this;
+    }
+
+    public function next(ServerRequestInterface $request, ResponseInterface $response, stdClass $args): ResponseInterface
+    {
+        $handler = $this->middleware[$this->middlewarePosition++] ?? $this->postware[$this->postwarePosition++] ?? null;
+
+        if ($handler === null) {
+            return $response;
+        }
+
+        return call_user_func("$handler::handle", $request, $response, $args, $this);
+    }
+}
```

Comment:
File should have a unix line ending it can cause very weird issues.

Just use https://marketplace.visualstudio.com/items?itemName=jkiviluoto.tws you can use this to add them automatically. It also removes trailing whitespaces